src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -4429,6 +4429,20 @@ public void run()
                     {
                         logger.info("failed to shutdown message service: {}", ioe);
                     }
+
+                    // wait for miscellaneous tasks like sstable and commitlog segment deletion
+                    ScheduledExecutors.nonPeriodicTasks.shutdown();
+                    try
+                    {
+                        if (!ScheduledExecutors.nonPeriodicTasks.awaitTermination(1, MINUTES))
```

Comment:
I agree that correctly order the shutdown sequence is desired. 

A minor concern is that 1 minute is not enough to wait for scheduled tasks to finish. For instance, the schema migration is scheduled 1 minute later in the future. So the migration task might again kick off after the await duration is over. 

Maybe we can also add such chunk of code in the `org.apache.cassandra.schema.MigrationCoordinator#pullSchema` method? To avoid initiation the schema pull. 

```java
        if (StorageService.instance.isLeaving())
        {
            logger.info("Skipped pulling schema from other nodes: the node is decommissing.");
            callback.fail();
            return;
        }
```

Reply:
I looked at executeExistingDelayedTasksAfterShutdown as suggested by @maedhroz and I think that one should work, I am pushing a commit now, let me know wdyt :-)